### PR TITLE
remove socket.io

### PR DESCRIPTION
**Pull request by memark**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

Related to https://github.com/socketio/socket.io/issues/3900

### CODE REVIEW

**Comment by murbanowicz** [comment]
@memark - The CI is failing: https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=96763&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=c264d665-0b79-59b7-7e04-9965cbe74860

Please fix it.

**Review by alexanderson1993 — APPROVED** [review]

**Comment by murbanowicz** [comment]
@memark unless GH is fooling me the CI is still failing

**Comment by memark** [comment]
@murbanowicz You are correct. Just sent the PR for DT-tools.

**Comment by memark** [comment]
The tests just started failing immediately locally with message
> AssertionError [ERR_ASSERTION]: The specified version 3.2.1 of just-clone must be newer than the version
it is supposed to replace, 3.2.2 of @types/just-clone.

(unrelated to my changes)

so I guess I'll need to push a few times here until the builds pass.

**Review by murbanowicz — APPROVED** [review]

**Comment by SimonSchick** [comment]
Why are you creating a dependency on socket.io in hapi, iirc it's not used there, I'd like more eyes on this.

File: types/hapi/package.json
Comment (SimonSchick) [inline types/hapi/package.json:5]: This appears to be only used as a test dependency, please remove/replace it.

File: types/hapi/v16/package.json
Comment (SimonSchick) [inline types/hapi/v16/package.json:5]: See above.

File: types/hapi__hapi/package.json
Comment (SimonSchick) [inline types/hapi__hapi/package.json:8]: See above.

**Comment by memark** [comment]
@SimonSchick I believe it is

>import * as SocketIO from "socket.io";

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/hapi/test/server/server-listener.ts#L3

File: types/micro/package.json
Comment (SimonSchick) [inline types/micro/package.json:4]: See above, this is used for testing only and should not be made a hard, consider replacing socket.io from the test as a whole.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

**Comment by SimonSchick** [comment]
Yes it is used in tests, but should not be shipped as a dependency of a package, either make it a dev dep or replace the test depency (it's not required most of the time).

File: types/micro/package.json
Comment (memark) [reply types/micro/package.json:4]: Thanks for your comment. The way I've done it is the way suggested in the README.

Are you the maintainer of the hapi types? Feel free to suggest how to get rid of this test dependency. The tests are very complicated to me, I'm not sure how I would go about even starting to change them.

Comment (memark) [reply types/micro/package.json:4]: Also, the test code is an exact copy from the hapi README:
https://github.com/hapijs/hapi/blob/master/API.md#-serverlistener

**Comment by memark** [comment]
@SimonSchick Adding it as a dev dep did not work
>Error: /home/vsts/work/1/s/types/hapi/package.json should not include field devDependencies

Should we delete that small test (server-listener) altogether then? Since it appears to test the integration with socket.io, it wouldn't make much sense exchanging it for something else.

**Comment by saboya** [comment]
There's no sense in putting that dependency in, not even as 'dev'. That excerpt in hapi's documentation is only an example, socket.io does not have any user visibility in hapi's API.

**Comment by memark** [comment]
@murbanowicz I need some advice here please. Turns out one of the tests for hapi is no longer relevant.

I tried deleting the test file (and updating tsconfig.json), that failed the build
>Error: Please delete all files in hapi when adding it to notNeededPackages.json.

I tried instead emptying the file but leaving it, that failed the build
>ERROR: 1:1  no-useless-files  File has no content.

What's the recommended way for this?

**Comment by murbanowicz** [comment]
@memark Unfortunately this is sort of an issue that I don't know how to deal with :/ 
You can try to ping the main repo maintainers and I'd expect them to give some advice here.

**Comment by memark** [comment]
@peterblazejewicz @sandersn Do you have any input on my question (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52719#issuecomment-834545013) how to remove a test file?

**Comment by peterblazejewicz** [comment]
Why are you rising TS minimum version to 3.8? (is this Socket.IO change? is that minimum TS version imposed now by Socket IO after migration to TS?). Some of those libs referenced here are just at v2 of the socket when checking their `package.json` (e.g. Cote: https://github.com/dashersw/cote/blob/master/package.json#L39)

**Comment by peterblazejewicz** [comment]
> Yes, you are correct, Cote is still using Socket 2. What's your advice here?

That's a problem. Same scenario like with Webpack 4/5. Let me check this scenario (Socket is to be removed completely, Webpack 5 was settled by using a passthrough types. I'll ask on Discord TS channel, in case you want to join discussion, thx!

**Comment by peterblazejewicz** [comment]
@memark  here: https://discord.com/channels/508357248330760243/757992230077333655/841570670270087168
I've asked about scenario with @types/socket.io beeing added to 'allowed' external dependencies. This would allow to avoid those tricks with using higher versions in `package.json` details (AFAIK, let's wait for the answer), thx!

**Comment by memark** [comment]
@peterblazejewicz Thank you for your efforts. Was there any answer?

I'm not able to join the DS channel. It just says
image

**Comment by sandersn** [comment]
A few things to point out. I may write more later:

1. @peterblazejewicz points out that socket.io dependents that haven't updated for socket.io@3 should still depend on `@types/socket.io`. In fact, sticking with `@types/socket.io` means that you don't otherwise have to update dependents. `@types/socket.io` has enough dependents to make this exception worthwhile, same as webpack.
2. The current CI failure happens because of the way that the tests compare to `master`. A new package named bogon exists in master but not in this branch, so it thinks you have deleted it. Just merge from master to get rid of the error.
3. I reproduced the bogus "please delete all of hapi" error locally, which happens because of deleting server-listener.ts. It's a bug in the test package which I'll have to fix.

**Comment by sandersn** [comment]
I just hit publish on a new release of the tests: https://github.com/microsoft/DefinitelyTyped-tools/releases/tag/v0.0.82 
So next time CI runs it should not say "please delete all of hapi" incorrectly.

**Comment by memark** [comment]
@sandersn @peterblazejewicz Thanks for your input.
Reading your comments I take it I should open a PR to add `@types/socket.io` to the allowed dependencies file, and then just atddv2 to existing dependant packages?
That seems a looooot easier. (Too bad that alternative wasn't mentioned in the readme. 😝 )

The tooling bug seems to have been fixed. Thanks!

**Comment by peterblazejewicz** [comment]
@memark yes, my intent, but wait for @sandersn to confirm :)

**Comment by sandersn** [comment]
Yep, that's right. The reason it's not in the readme is that socket.io is only the second big deprecation we've done, where the deprecated package leaves behind a lot of dependent packages like plugins. 2 times might be enough to add it to the README though.

**Comment by memark** [comment]
@sandersn @peterblazejewicz

...and finally the CI passes!!! 🎉

**Comment by andrewbranch** [comment]
@memark it looks like you accidentally committed a change to the permission mode of types/1line-aa/tslint.json 🤔 

image

Undo that and then we can get this thing merged 🚂

**Comment by memark** [comment]
@andrewbranch Well, that was weird... Nice catch, restored it now!

**Review by andrewbranch — APPROVED** [review]

**Comment by memark** [comment]
Ready to merge

**Comment by peterblazejewicz** [comment]
@memark thx for your patience!

**Comment by floriancargoet** [comment]
Since this change, I've conflicting types.
```
    // Session is shared between Express & socket.io.
    io.use(
      sharedSession(this.sessionMiddleware, {
        autoSave: true,
      })
    );
```
The error is `Types of parameters 'socket' and 'socket' are incompatible.`.
`io.use()` uses the `Socket` from `socket.io` and `sharedSession()` uses the `Socket` from `@types/socket.io`.

**Comment by memark** [comment]
@floriancargoet It's not clear to me what packages and what versions of those packages you are using. Could you provide a full repro?

**Comment by floriancargoet** [comment]
I'm sorry, my comment was really useless (no repro, no versions…).
I've tried to reproduce it but couldn't after deleting node_modules so I guess something was wrong there.

I'm sorry for wasting your time @memark.

**Comment by memark** [comment]
No worries at all. Glad it works for you now!

**Pull request merged**
